hadoop-hdds/docs/content/design/find-block-missing-key.md

```diff
@@ -0,0 +1,106 @@
+    - After collecting sufficient Key metadata, organize it for further processing.
+    - There are three approaches to retrieve metadata (detailed in [Three Approaches to Retrieve Metadata](#three-approaches-to-retrieve-metadata)).
+2. Organize Keys by Container:
+    - Group Keys by their respective Containers in local memory. Keys within the same Container are grouped together, and relevant information is recorded.
+3. Query Container Status:
+    - Send requests to the SCM (Storage Container Manager) to obtain Container locations (i.e., which DNs the Container resides on).
+    - If the Container is in a `DELETED` or `DELETING` state, or if the Container cannot be found, the Key is marked as a Block Missing Key.
+    - If the Container is found and its status is normal, continue to check the existence of the Blocks.
+    - Containers in the `OPEN` and `RECOVERING` state should be skipped. However, for Containers in the `CLOSING` or `QUASI_CLOSED` state, further investigation is needed, as these states may indicate Block issues.
+4. Query Block Existence:
+    - Based on the information provided by the SCM, send `headBlocks` requests to the DNs to verify the existence of the Blocks.
+    - A single `headBlocks` request can query multiple Blocks at once.
+    - A new DN API for `headBlocks` requests needs to be developed.
+5. Determine Block Missing Keys:
+    - Once the existence of all Block replicas for a Key is verified, the client determines whether the Key is a Block Missing Key based on its replication tolerance:
+        - For simple replication: the total number of replicas.
+        - For EC Keys: Parity Cell count + 1.
+6. Final Verification:
+    - Perform an additional verification for detected Block Missing Keys to confirm whether the Key has been deleted.
+    - If the Key is no longer present in OM, it is confirmed as a Block Missing Key.
+
+
+## Additional Features
+
+1. Result Output: Supports exporting results to a specified file in a predefined format.
+2. Progress Tracking: Supports saving scan progress to enable resumption from checkpoints.
+3. Partial Replica Check: Can detect Keys with partially missing replicas. This feature may overlap with the `find-missing-padding` command.
+4. Parallel Bucket Scanning: Supports scanning multiple buckets concurrently.
+5. Retry Mechanism: For large clusters, where scans may take significant time, and DN restarts are likely, the tool should support retries to improve reliability.
+
+
+## <a name="three-approaches-to-retrieve-metadata"></a>Three Approaches to Retrieve Metadata
+
+1. Using `listStatusLight` and `listKeysLight`:
+    - Advantages: Lightweight and high performance.
+    - Disadvantages: Requires API enhancements to include Block information.
+2. Using `listStatus` and `listKeys`:
+    - Advantages: Can be used directly without additional development.
+    - Disadvantages: Lower performance and may become a bottleneck.
```

Comment:
The lower performance comes form the block information that is passed in. If we use `listKeysList` and add block information, it will be the same as `listKeys`.

Reply:
Probably just add the BlockID (ContaienrID + LocalID) which should have less of an impact.